support generic resource action invoke

williexu opened the pull request:
---
Fixes https://github.com/Azure/azure-cli/issues/2607
Samples (on vm resource):
`az resource invoke-action --action powerOff --ids <vm_id>`
`az resource invoke-action --action capture --ids <vm_id> --parameters {"vhdPrefix":"myPrefix","destinationContainerName":"container","overwriteVhds":"true"}`
`az resource invoke-action --action capture --ids <vm_id> --parameters @{file}`




This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [x] Each command and parameter has a meaningful description.
- [x] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

msftclas commented:
@williexu,
Thanks for your contribution as a Microsoft full-time employee or intern. You do not need to sign a CLA.
_Thanks,
Microsoft Pull Request Bot_

yugangw-msft commented on src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py:
```diff
@@ -8,68 +8,59 @@
-    short-summary: Create a managed application.
+    short-summary: Creates a managed application.
     examples:
-        - name: Create a managed application of kind 'ServiceCatalog'. This requires a valid managed application definition ID.
```
Please rebase to get rid of unrelated changes

yugangw-msft commented on src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py:
```diff
@@ -41,6 +40,9 @@
 register_cli_argument('resource move', 'ids', nargs='+')
+register_cli_argument('resource invoke-action', 'action', help='The action that will be invoked')
+register_cli_argument('resource invoke-action', 'parameters', help='Parameter arguments for the action that will be passed along in the post request body')
+register_cli_argument('resource invoke-action', 'post_url', help='The url that will be sent in the POST request (must include api-version query parameter)')
```
Per on going mail thread, we might expose a general request url which you can apply any http method

yugangw-msft replied:
Sorry for changing the mind, but let us get rid of it.
We can move such functionalities into ms-rest-azure moving forward

yugangw-msft commented on src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py:
```diff
@@ -520,6 +520,23 @@ def tag_resource(tags,
+                           resource_id=None, api_version=None):
+    """ Invokes the provided action on an existing resource. Action specified through either action or post-url"""
+    if post_url:
+        if action or resource_group_name or resource_id or resource_name or resource_type:
```
suggest: `any([action, resource_group_name, resource_id, resource_name, resource_type])`

yugangw-msft commented on src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py:
```diff
@@ -1359,6 +1376,70 @@ def tag(self, tags):
+        """
+        Formats Url if none provided and sends the POST request with the url and parameters.
+        """
+        from msrestazure.azure_operation import AzureOperationPoller
```
//CC: @lmazuel, we are accessing private fields like following. Feel free to comment if we need adjustments to get things maintainable. We are having e2e test coverage, so should be fine

yugangw-msft commented on src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py:
```diff
@@ -1359,6 +1376,70 @@ def tag(self, tags):
+
+        def get_long_running_output(response):
+            from msrestazure.azure_exceptions import CloudError
+            if response.status_code not in [200, 202, 204]:
```
If we will cover non-post action, let us cross check to make sure we include more valid http return codes

troydai replied:
Shouldn't all status code between 200 and 299 acceptable?

lmazuel replied:
According to ARM this is enough:
https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#call-action-post

davidebbo commented:
What about actions that take a body, like `slotsswap` on a `Microsoft.Web/sites` resource? Or is that what `--parameters` is for? Not super clear whether that refers to query string or request body.

williexu commented:
> What about actions that take a body, like slotsswap on a Microsoft.Web/sites resource? Or is that what --parameters is for? Not super clear whether that refers to query string or request body.

@davidebbo , Yes, that is what `--parameters` is for. It allows the user to specify any JSON to be sent along in the request body. See the following example:
 `az resource invoke-action --action capture --ids <vm_id> --parameters {"vhdPrefix":"myPrefix","destinationContainerName":"container","overwriteVhds":"true"}`. 
Sorry if this wasn't more clear.

yugangw-msft commented:
@williexu, @davidebbo 
Suggest to rename `--parameters` to `--request-body`? We don't need to name the exactly same like PS. `--parameters` is a too general to me

williexu commented:
Addressed comments. I have removed the `--post-url` parameter as per discussion with @yugangw-msft . This was included as a way to allow users with niche requests to specify their own url to post. This will instead be considered for ms-rest-azure if needed. 

@davidebbo I changed the name of `--parameters` to `--request-body` instead. Hopefullt this will make the argument more clear.

yugangw-msft commented on src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/tests/test_resource.py:
```diff
@@ -749,5 +749,28 @@ def test_group_deployment_crossrg(self, resource_group, resource_group_cross):
+
+        vm_id = vm_json.get('id', None)
+
+        self.cmd('resource invoke-action --action powerOff --ids {}'.format(vm_id))
```
can you add a `resource show` to verify the machine was indeed powered off?

yugangw-msft commented on src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py:
```diff
@@ -1266,7 +1277,7 @@ def __init__(self,
         self.resource_group_name = resource_group_name
         self.resource_provider_namespace = resource_provider_namespace
-        self.parent_resource_path = parent_resource_path
+        self.parent_resource_path = parent_resource_path if parent_resource_path else ''
```
if you like to do this, then you can remove all occurrences of `self.parent_resource_path or ''` in the same class
Or you back the change out and apply ` self.parent_resource_path or ''` inside `invoke_action`. Either way, let us ensure the code is consistent.

davidebbo commented:
Sounds, good!

yugangw-msft commented on src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py:
```diff
@@ -1359,6 +1370,69 @@ def tag(self, tags):
 
+    def invoke_action(self, action, request_body):
+        """
+        Formats Url if none provided and sends the POST request with the url and parameters.
```
rename `parameters` to `request_body`

troydai commented on src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py:
```diff
@@ -41,6 +40,8 @@
 
 register_cli_argument('resource list', 'name', resource_name_type)
 register_cli_argument('resource move', 'ids', nargs='+')
+register_cli_argument('resource invoke-action', 'action', help='The action that will be invoked')
```
What action? Http action? Should we elaborate in the help?

tjprescott replied:
How would someone know what actions are possible on a given resource?

troydai commented on src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py:
```diff
@@ -1359,6 +1370,69 @@ def tag(self, tags):
+
+        url = '/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/' \
+            '{resourceProviderNamespace}/{parentResourcePath}/{resourceType}/{resourceName}/{action}'
+        path_format_arguments = {
```
No need to pass keyword parameter in this form.

troydai commented on src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py:
```diff
@@ -1359,6 +1370,69 @@ def tag(self, tags):
+        header_parameters = {}
+        header_parameters['Content-Type'] = 'application/json; charset=utf-8'
+        if self.rcf.resources.config.generate_client_request_id:
+            header_parameters['x-ms-client-request-id'] = str(uuid.uuid1())
```
Using `uuid4`

tjprescott commented on src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py:
```diff
@@ -326,7 +326,7 @@
 helps['resource invoke-action'] = """
     type: command
-    short-summary: Invoke an action on the resource. Must provide either the --action parameter or the --post-url parameter
+    short-summary: Invoke an action on the resource. Must provide the --action parameter
```
> Must provide the --action parameter 

In that case the parameter should be [Required] and you don't need to state that here.

tjprescott commented on src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py:
```diff
@@ -40,7 +40,7 @@
 register_cli_argument('resource list', 'name', resource_name_type)
 register_cli_argument('resource move', 'ids', nargs='+')
-register_cli_argument('resource invoke-action', 'action', help='The action that will be invoked')
+register_cli_argument('resource invoke-action', 'action', help='The action that will be invoked on the specified resource')
```
That doesn't really make it much more clear. How will someone know what the possible actions are for a resource?  Is there another endpoint that documents this?

yugangw-msft replied:
This is a generic command as a stop gap where there is no strong type CLI command for that. So it targets to advance users or who are suggested by such people. Under that context, people are expected to check out the rest api themselves which are well documented in msdn, like
https://docs.microsoft.com/en-us/rest/api/compute/virtualmachines/virtualmachines-start.
We should call out such sources in the help though.

leidegre commented:
Also `--request-body` should be able to be specified as a file, a command line on Windows can at most be around 8000 characters, that's not that much if you have a big resource with lots of things in it.

williexu commented:
@leidegre, yes it is possible and probably preferred to specify a filepath for the request body. Use @{file} to utilize this option

leidegre commented:
@williexu and if I just want to PUT a resource, can I do that to? PUT this resource from this file into Azure? Like `az resource update` but with a file as source.

tjprescott commented:
@leidegre I believe the command you are looking for is `az resource create`.

leidegre commented:
@tjprescott I believe you just made my day.

lmazuel commented on src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py:
```diff
@@ -1353,12 +1364,75 @@ def tag(self, tags):
+                exp = CloudError(response)
+                exp.request_id = response.headers.get('x-ms-request-id')
+                raise exp
+            return deserialize('str', response)
```
`return response.text` would be faster
